### ISSUE TITLE
Low Error - But Low Quality Tet Mesh

### ISSUE THREAD

**Issue by 9of9** [NVlabs/nvdiffrec#37]
I've run a few tests with some in-the-wild examples from my photogrammetry collection, and I seem to be consistently running into the same issue: the combined output converges to something that is very, very similar to the target images, but the quality of actual tet mesh is garbage compared to what you get from the nerd data.

For instance, from 19 images:
![img_mesh_pass_000049](https://user-images.githubusercontent.com/31524061/170835797-e0035355-44e9-46d6-bcd4-b90a1651e910.png)

The rendered output clearly converges pretty close to the desired result and kd seems maybe close-ish to what you'd want, but the probe, ks and n don't converge at all on what you would expect as correct values. The resulting tet mesh holds roughly the right shape, but generally just looks like swiss cheese:

![image](https://user-images.githubusercontent.com/31524061/170836116-71c5a900-84ac-43c8-a45a-43034d288cd4.png)

Another example from 45 images:
![image](https://user-images.githubusercontent.com/31524061/170836199-f913f9f9-6198-425c-a136-edfcf35bb1f0.png)

![image](https://user-images.githubusercontent.com/31524061/170836374-bd69eb06-a437-4846-bc45-3e56c75462c0.png)

This one generated a better mesh, but seemed to over-optimise ks for a surface which should largely be very diffuse. On the inside, the mesh is still very much swiss cheese.

This seems in contrast to very solid results on synthetic data. Do you have any recommendations for how to tune the training for better results? I have used similar configs to what are used for the nerd datasets, but I'm not sure if there are better ways of controlling the learning rates. Or is the trick ultimately that it just needs more photos?

**Comment by JHnvidia** [comment]
Hi @9of9,

I'm just speculating, but 19 views sounds like it's probably on the low side. Additionally you can expect mesh quality to degrade if camera poses and segmentation masks are inaccurate. 

Since we're doing a full SDF grid for DMTet (as supposed to some smooth representation like an MLP) it is prone to fail towards a high frequency "mess". It's a tradeoff we had to make because overly smoothing geometry doesn't work that well with very complex scenes (e.g. Lego from the nerf dataset). It is possible to regularize geometry by using different reprensentations, and we did that for the DTU datasets in the paper, but unfortunately the code is too out of date & ad hoc to be released.

The specular term is hardest to train (and where most of our issues lie). If you know that your scenes only contain diffuse materials you can restrict the specular terms from the config. They are: [unused, roughness, metalness], so the following should force the optimizer to only consider rough diffuse-like surfaces.
```
    "ks_min" : [0, 1, 0],
    "ks_max" : [0, 1, 0],
```

You could also change to a diffuse BSDF by chaining the following line to `= 'diffuse'`, but I havent tried that code path so I'm not 100% sure if it works.
https://github.com/NVlabs/nvdiffrec/blob/main/train.py#L176

I would also suggest comparing the results of both passes, because sometimes the second pass can degrade geometric quality (but this doesn't seem to be the cause in your scenes). There's a ` "lock_pos" : true,` flag to avoid moving the geometry in the second pass.

**Comment by 9of9** [comment]
Thank you, that's very thorough! I will investigate and see if there's a way to improve the results :)

**Comment by tlok666** [comment]
> Thank you, that's very thorough! I will investigate and see if there's a way to improve the results :)

Hi, I come across the same problem. Did you solve it by adjusting the ks values?

**Comment by 9of9** [comment]
> > Thank you, that's very thorough! I will investigate and see if there's a way to improve the results :)
> 
> Hi, I come across the same problem. Did you solve it by adjusting the ks values?

It was better in some respects, but not dramatically. Here I constrained the lego house to be non-metallic and relatively shiny:
![image](https://user-images.githubusercontent.com/31524061/173845292-dc8ae3ed-bf2c-4cb4-be42-8d06c97342b2.png)

And same with the tower above that I constrained to be fairly rough and non-metallic:
![image](https://user-images.githubusercontent.com/31524061/173845448-5295d534-0c7b-47bf-b23d-09d47e9b17e6.png)

The results are a little more sensible and conventional, it has made the diffuse in particular behave better, and the env maps generated from these are a little bit more plausible, but the quality of the tet meshes is still quite poor. I suspect the only real solution will be to use many more source images.

**Comment by tlok666** [comment]
![image](https://user-images.githubusercontent.com/12291907/173974302-dbb2f0c0-afc3-4c61-b8af-d9afc3d52ade.png)
In my case of human mesh, although I have adjusted the ks values to [0, 1, 0].  It seems like a low error but low-quality mesh.

**Comment by tlok666** [comment]
> > > Thank you, that's very thorough! I will investigate and see if there's a way to improve the results :)
> > 
> > 
> > Hi, I come across the same problem. Did you solve it by adjusting the ks values?
> 
> It was better in some respects, but not dramatically. Here I constrained the lego house to be non-metallic and relatively shiny: ![image](https://user-images.githubusercontent.com/31524061/173845292-dc8ae3ed-bf2c-4cb4-be42-8d06c97342b2.png)
> 
> And same with the tower above that I constrained to be fairly rough and non-metallic: ![image](https://user-images.githubusercontent.com/31524061/173845448-5295d534-0c7b-47bf-b23d-09d47e9b17e6.png)
> 
> The results are a little more sensible and conventional, it has made the diffuse in particular behave better, and the env maps generated from these are a little bit more plausible, but the quality of the tet meshes is still quite poor. I suspect the only real solution will be to use many more source images.

It seems like, nvdiffrec could generate limited mesh quality in complicated cases. A related paper refers to the same issues of 'Low Error - But Low Quality Tet Mesh'. (https://arxiv.org/pdf/2206.03380.pdf)
![image](https://user-images.githubusercontent.com/12291907/174926562-547a279f-2959-47fd-9946-7afbf5c4b737.png)

**Comment by YLongJin** [comment]
> lock_pos

Hello, bro 
I got the same issues as you( Low Quality Tet Mesh)，but  my result of img_dmtest is not better than you. And tet mesh is like a shit!   Could you plz tell me
